### ISSUE TITLE
Feature request: avoid repeated unpacking of exactly the same app

### ISSUE THREAD

**Issue by bersbersbers** [pyinstaller/pyinstaller#7907]
**Is your feature request related to a problem? Please describe.**
`onefile` mode is great for distribution. No `zip` files, no installer, just copy an `exe` and run. However, `onefile` is slow, because every time it runs it extracts the exact same data into almost the same location. In my case, we are speaking 4.5 vs. 1.5 seconds launch times (hand-stopped).

**Describe the solution you'd like**
I would find it tremendously useful if I could opt-in to a process that works somewhat like this:
- During generation of the `onefile` bundle, define an identifier for the app and generate a unique ID (maybe a hash of the content) that can be quickly accessed at run-time
- During run-time, check if `$TEMP\_MEI_$APP\$UNIQUE` exists
  - yes: launch app from that location without extracting further
  - no: extract app into that location and run from there
- On close, remove other/older folders in `$TEMP\_MEI_$APP` that are not `$UNIQUE`.

One could name this hybrid mode, or persistent onefile mode, or something that effect.

**Comment by rokm** [comment]
This comes up every now and then (see for example #7879), but as far as I'm concerned, the answer remains no. The bootloader code is complicated enough as it is.

**Comment by bersbersbers** [comment]
> This comes up every now and then

Maybe that's a sign :) I'll add to the discussion in #7879.

**Comment by bwoodsend** [comment]
> On close, remove other/older folders in $TEMP\_MEI_$APP that are not $UNIQUE.

That would mean that if you have two unrelated applications built by PyInstaller, they'd be forever deleting each other's unpacked directories -- possibly even whilst the other application is still running.

At a slightly more persnickety level, there's also the joys of concurrency to deal with -- launch the app twice in quick succession so that the second instance sees the partially unpacked directory, skips its own unpacking and crashes because the extraction isn't finished yet. And what would happen if the application was aborted midway through unpacking -- `$TEMP\_MEI_$APP\$UNIQUE` would always exist so the next run would always skip the unpack and crash every time it launches.

Also, bear in mind that this all has to be written in C. All this filtering will have to be written from scratch. It has to be platform and locale agnostic, unicode safe, multi-user safe and combed through for anything that can be tricked into modifying or executing the wrong file. The cost of this feature is a lot higher than you seem to realise.

**Comment by bersbersbers** [comment]
> > On close, remove other/older folders in $TEMP_MEI_$APP that are not $UNIQUE.
> 
> That would mean that if you have two unrelated applications built by PyInstaller, they'd be forever deleting each other's unpacked directories [...]

... *if* those two unrelated applications share the same `$APP` identifier, yes: $APP needs to be sufficiently unique. That may still happen with different version of the same app, though.

> possibly even whilst the other application is still running

That is impossible on standard Windows, but you are right, it's a risk on other operating/file systems.

> At a slightly more persnickety level, there's also the joys of concurrency to deal with -- launch the app twice in quick succession so that the second instance sees the partially unpacked directory, skips its own unpacking and crashes because the extraction isn't finished yet. And what would happen if the application was aborted midway through unpacking -- `$TEMP\_MEI_$APP\$UNIQUE` would always exist so the next run would always skip the unpack and crash every time it launches.

Good points. The extracting application would have to mark the target folder as "complete", e.g., by placing a file inside the folder, once it is done extracting. The following application would have to poll the existing folder for said file, and if that is not created, (prompt the user to) delete the folder after a timeout.

> Also, bear in mind that this all has to be written in C. [...] The cost of this feature is a lot higher than you seem to realise.

This being in C is in fact a point I was not aware of. Again, benefit-cost-ratio is a reason that I can easily accept, and this being C underlines that point.

Is does sound like the way to go is using onedir mode and a good SFX library - from your experience with pyinstaller, can you recommend one by any chance? :)

**Comment by bwoodsend** [comment]
> Is does sound like the way to go is using onedir mode and a good SFX library - from your experience with pyinstaller, can you recommend one by any chance? :)

You'll have to expand the *SFX* acronym for me. The only meaning I know for it is sound effects.

**Comment by bersbersbers** [comment]
Sorry! https://en.wikipedia.org/wiki/SFX:
> [Self-extracting archive](https://en.wikipedia.org/wiki/Self-extracting_archive), a compressed file with an embedded executable to decompress itself

(I must admit I believe that acronym has lost popularity over the decades.)

**Comment by bwoodsend** [comment]
I've never actually used one. I'm quite fond of 7zip though so I'd probably go with that if I had to pick without trying.

Honestly, I never understood why people don't like zip files. It's just *right click* -> *extract here* and you're done.

**Comment by bersbersbers** [comment]
> Honestly, I never understood why people don't like zip files.

Thanks! I think there's at least two reasons. A historic one is that earlier versions of Windows came without unzipping capabilities. Another one is that if you build on exe on CI, you can download *and run* it without leaving the browser. Compare https://gitlab.com/gitlab-org/gitlab/-/issues/37384, for instance.

**Comment by bwoodsend** [comment]
> A historic one is that earlier versions of Windows came without unzipping capabilities.

I've just noticed that Windows 7 comes with zip file support. The oldest version of Python that PyInstaller supports is 3.8 and Python hasn't supported anything older than Windows 7 since Python 3.4 so unless you downgrade everything about 4 years worth of releases, you shouldn't be able to support these legacy Windows versions anyway.

**Comment by bersbersbers** [comment]
Yes, that's why I wrote "historic" :) sometimes, however, preferences seem to take a while to change.

**Comment by pa-0** [comment]
While I understand and respect the views / vision of the author/contributors to the Project, I think that there is still a use case for this that aligns with that vision.  As mentioned in an explanation given by a PyInstaller contributor in a previous issue:

> [**@nchammas**](https://github.com/nchammas) commented [on Jan 20, 2016](https://github.com/pyinstaller/pyinstaller/issues/1782#issuecomment-173221712)
 
>... PyInstaller's job is to bundle up everything a Python application needs into a single package so it runs without requiring any external Python dependencies. Its job is not to make modifications to the system.
> 
> It's up to the user to download the package they want, and it's up to the user to place it in the location they want. PyInstaller doesn't place any binaries on the user's PATH, or copy files to permanent locations on the user's system. That's why when you run a --onefile executable, everything that PyInstaller does outside of the package gets cleaned up as soon as the executable terminates.
.."

What @nchammas describes here sounds like the behavior of a typical standalone, **portable** executable, or at least, the kind of features/behavior that users are looking for in a portable app ("...run without requiring external python dependencies," "not...make modifications to the system," "download the package <ins><strong>where they want</strong></ins>").  Considering these behaviors, with special attention especially given to offering the user the ability to download it where they want, it makes sense that the portable would limit its footprint to the directory where it is downloaded.  There are plenty of scenarios where users prefer that anything saved to the system by the executable (at least where foreseeable and to a reasonable degree controllable) is contained to the folder where they place that executable, if even just for the sake of visibility and assurance that the directory is removed once the associated process are terminated as mentioned in the cited reference above. 

Consider the user that places the executable on a flash drive, expecting it to remain entirely on that external drive, when they run it on any number of systems.  I think in these situations, at least having the option to prevent the executable from extracting to a less easily accessible / knowable folder in the hidden temp directory, aligns with this idea that the user gets to decide where to place the executable, etc.

To further extrapolate on this logic, if the use case is a portable exe that is confined to the directory where it is placed and the contents of the executable are extracted to a location completely transparent to the user (such as a folder on their flash drive), it makes sense to refrain from deleting the extracted contents (or at least have the option to refrain), if doing so would improve the user's experience running the app.

**Comment by bwoodsend** [comment]
If you don't want to keep the extracted application out of the user's sight/way then why wouldn't you just put a onedir bundle into an archive?

**Comment by emilebosch** [comment]
Hey all, just ran into this. Are there any forks that have this capability? Seems like a simple no-brainer to me. Regardless of historic reasons, I don't want my users to wait and waste extra CPU cycles. @bersbersbers 


You can just put it in a `~.pyinstaller` cache directory or something and fallback to tmp if its impossible to do so

**Comment by KebabLord** [comment]
I've been serving my software with --onefile flag since 2019.
I handle the binary downloads externally upon first time progam launch, i download them and save them to a dedicated folder of software in `~/Documents/my_app`. Small binary size, faster launch. I don't think this feature is an essential if it's an headache to implement.

**Comment by xjcl** [comment]
I think `cx_Freeze` does exactly this, doesn't it?
